pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,166 +18,480 @@ import (
 	"context"
 	"fmt"
 	"math"
-	"strings"
 
-	"github.com/Pallinder/go-randomdata"
-	"github.com/mitchellh/hashstructure/v2"
+	"github.com/aws/karpenter/pkg/utils/apiobject"
+
+	"go.uber.org/multierr"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
-	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
-	"github.com/aws/karpenter/pkg/utils/functional"
 	"github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/sets"
 )
 
 type Topology struct {
 	kubeClient client.Client
+	topologies map[uint64]*TopologyGroup
+	// Anti-affinity works both ways (if a zone has a pod foo with anti-affinity to a pod bar, we can't schedule bar to
+	// that zone, even though bar has no anti affinity terms on it. For this to work, we need to separately track the
+	// topologies of pods with anti-affinity terms, so we can prevent scheduling the pods they have anti-affinity to
+	// in some cases.
+	inverseTopologies map[uint64]*TopologyGroup
+	constraints       *v1alpha5.Constraints
```

Comment:
I don't think this is used anymore. I'm thinking of this topology object as a longlived tracker/manager, so constraints wouldn't make sense anyways. It should even be shared across provisioners.

Reply:
It's only used once place now which is when creating new topology groups.  We need to pull off the available zones/capacity types to initialize those.  If in the future it's reused across provisioners we could initialize with the union of all provisioner supported constraints.  I think we already tighten by provisioner constraints, so having extra domains there wouldn't hurt normally, it could get weird though:

Assume,
- provisioner 1 can schedule in zone-a, zone-b
- provisioner 2 can schedule in zone-a, zone-b, zone-c

If a deployment with zonal spread/max-skew=1 could only be provisioned by provisioner 1 due to a label seletor, we could run a maximum of four pods.  The topology would know about the zone-c and prevent scheduling more.  If this is correct behavior or not depends on your viewpoint :)

Reply:
Sounds correct to me